# Práctica 1

## Aforos de peatones en la ciudad de Madrid

En esta práctica trabajaremos con información relativa a la cantidad de peatones correspondientes a diferentes calles de Madrid en diferentes días y horas de la semana.

Se realizan mediciones en determinadas calles en puntos concretos con periodicidad diaria cada 15 minutos. 

NOTA INFORMATIVA.  Por haberse producido una avería técnica en el sistema de recogida de datos, detectada en el análisis de los datos del mes de Julio y Agosto de 2019 al reflejarse un descenso significativo en todas las Estaciones Permanentes tras una recalibración de los puntos de medida, no se disponenen de los datos de esos dos meses.

## Dataset

El fichero con el que vamos a trabajar se llama `'peatones_madrid.csv'`. Se trata de un fichero de texto cuyo separador de campos es el símbolo `|`.

In [64]:
###
## Nombre y apellidos: Mateo Arellano 
###

## Descripción del dataset

| Nombre | Descripción |
|:--------|:-------------|
| Fecha |  Fecha de recogida de datos | 
| Hora  |  Hora de recogida de datos
| Identificador |  Identificador del aforo. Indicador del número de estación Permanente de aforo | 
| Peatones        |  Número de peatones contadas en un cuarto de hora | 
| Número_Distrito   |  Número del distrito donde se encuentra la estación de aforo | 
| Distrito        |  Denominación del Distrito donde se encuentra la estación de aforo| 
| Nombre_Vial           | Calle donde se encuentra la estación de aforo | 
| Número                     | NÚMERO DE LA CALLE donde se encuentra la estación de aforo | 
| Código_Postal       | CÓDIGO POSTAL donde se encuentra la estación de aforo
| Latitud                   |  LATITUD en coordenadas geográficas | 
| Longitud                 |  LONTITUD en coordenadas geográficas | 

In [65]:
import pandas as pd
import numpy as np


In [66]:
peaton = pd.read_csv("peatones_madrid/peatones_madrid.csv",delimiter = "|")
peaton

Fecha     Hora    Identificador  Peatones  \
0                2019-01-01  0:00:00  PERM_PEA02_PM01     497.0   
1                2019-01-01  0:15:00  PERM_PEA02_PM01     244.0   
2                2019-01-01  0:30:00  PERM_PEA02_PM01     179.0   
3                2019-01-01  0:45:00  PERM_PEA02_PM01     174.0   
4                2019-01-01  1:00:00  PERM_PEA02_PM01     131.0   
...                     ...      ...              ...       ...   
504093  2020-12-31 23:00:00    23:00  PERM_PEA15_PM01      44.0   
504094  2020-12-31 23:00:00    23:00  PERM_PEA16_PM01      44.0   
504095  2020-12-31 23:00:00    23:00  PERM_PEA17_PM01     133.0   
504096  2020-12-31 23:00:00    23:00  PERM_PEA18_PM01     191.0   
504097  2020-12-31 23:00:00    23:00  PERM_PEA19_PM01      48.0   

        Número_Distrito     Distrito                   Nombre_Vial  Número  \
0                   1.0       Centro              Calle Fuencarral    22.0   
1                   1.0       Centro              Calle Fuencarral    22.0   
2                   1.0       Centro              Calle Fuencarral    22.0   
3                   1.0       Centro              Calle Fuencarral    22.0   
4                   1.0       Centro              Calle Fuencarral    22.0   
...                 ...          ...                           ...     ...   
504093              1.0       Centro              Calle Fuencarral    22.0   
504094              2.0   Arganzuela  Plaza del Emperador Carlos V    11.0   
504095              2.0   Arganzuela  Plaza del Emperador Carlos V    11.0   
504096              1.0       Centro              Calle Fuencarral    22.0   
504097              1.0       Centro              Calle Fuencarral    22.0   

        Código_Postal      Latitud      Longitud  
0             28004.0   40,4220090    -3,7008917  
1             28004.0   40,4220090    -3,7008917  
2             28004.0   40,4220090    -3,7008917  
3             28004.0   40,4220090    -3,7008917  
4             28004.0   40,4220090    -3,7008917  
...               ...          ...           ...  
504093        28004.0  40,42200905  -3,700891722  
504094        28012.0  40,41110888  -3,696471467  
504095        28012.0  40,41110888  -3,696471467  
504096        28004.0  40,42200905  -3,700891722  
504097        28004.0  40,42200905  -3,700891722  

[504098 rows x 11 columns]

## Cuestiones a resolver

### Características
Usa la función `info()` para mostrar las características principales del dataframe.

In [67]:
peaton.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 504098 entries, 0 to 504097
Data columns (total 11 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   Fecha            504098 non-null  object 
 1   Hora             504098 non-null  object 
 2   Identificador    504098 non-null  object 
 3   Peatones         504098 non-null  float64
 4   Número_Distrito  504098 non-null  float64
 5   Distrito         504098 non-null  object 
 6   Nombre_Vial      504098 non-null  object 
 7   Número           504098 non-null  float64
 8   Código_Postal    504098 non-null  float64
 9   Latitud          504098 non-null  object 
 10  Longitud         504098 non-null  object 
dtypes: float64(4), object(7)
memory usage: 42.3+ MB


### Primeros estadísticos

Usa la función `describe()` para calcular los primeros estadísticos del dataset.

In [68]:
peaton.describe()

Peatones  Número_Distrito         Número  Código_Postal
count  504098.000000    504098.000000  504098.000000  504098.000000
mean      444.895973         2.646132      25.708876   28009.812011
std       726.659314         2.995710      22.737269       4.421950
min        -1.000000         1.000000       1.000000   28001.000000
25%        78.000000         1.000000      10.000000   28004.000000
50%       252.000000         1.000000      22.000000   28012.000000
75%       603.000000         2.000000      36.000000   28013.000000
max     50460.000000        10.000000      95.000000   28015.000000

### Mostrar datos
Las funciones `head()`  y  `tail()` muestran las primeras 5 filas y las 5 últimas filas del dataframe respectivamente. Usalas para probar su funcionamiento.

In [69]:
peaton.head()

Fecha     Hora    Identificador  Peatones  Número_Distrito Distrito  \
0  2019-01-01  0:00:00  PERM_PEA02_PM01     497.0              1.0   Centro   
1  2019-01-01  0:15:00  PERM_PEA02_PM01     244.0              1.0   Centro   
2  2019-01-01  0:30:00  PERM_PEA02_PM01     179.0              1.0   Centro   
3  2019-01-01  0:45:00  PERM_PEA02_PM01     174.0              1.0   Centro   
4  2019-01-01  1:00:00  PERM_PEA02_PM01     131.0              1.0   Centro   

        Nombre_Vial  Número  Código_Postal     Latitud    Longitud  
0  Calle Fuencarral    22.0        28004.0  40,4220090  -3,7008917  
1  Calle Fuencarral    22.0        28004.0  40,4220090  -3,7008917  
2  Calle Fuencarral    22.0        28004.0  40,4220090  -3,7008917  
3  Calle Fuencarral    22.0        28004.0  40,4220090  -3,7008917  
4  Calle Fuencarral    22.0        28004.0  40,4220090  -3,7008917

In [70]:
peaton.tail()

Fecha   Hora    Identificador  Peatones  \
504093  2020-12-31 23:00:00  23:00  PERM_PEA15_PM01      44.0   
504094  2020-12-31 23:00:00  23:00  PERM_PEA16_PM01      44.0   
504095  2020-12-31 23:00:00  23:00  PERM_PEA17_PM01     133.0   
504096  2020-12-31 23:00:00  23:00  PERM_PEA18_PM01     191.0   
504097  2020-12-31 23:00:00  23:00  PERM_PEA19_PM01      48.0   

        Número_Distrito     Distrito                   Nombre_Vial  Número  \
504093              1.0       Centro              Calle Fuencarral    22.0   
504094              2.0   Arganzuela  Plaza del Emperador Carlos V    11.0   
504095              2.0   Arganzuela  Plaza del Emperador Carlos V    11.0   
504096              1.0       Centro              Calle Fuencarral    22.0   
504097              1.0       Centro              Calle Fuencarral    22.0   

        Código_Postal      Latitud      Longitud  
504093        28004.0  40,42200905  -3,700891722  
504094        28012.0  40,41110888  -3,696471467  
504095        28012.0  40,41110888  -3,696471467  
504096        28004.0  40,42200905  -3,700891722  
504097        28004.0  40,42200905  -3,700891722

### Selección 1

Selecciona la columna `Fecha` del dataset. ¿De qué tipo es la columna `Fecha`? . 

In [71]:
peaton.Fecha

0                  2019-01-01
1                  2019-01-01
2                  2019-01-01
3                  2019-01-01
4                  2019-01-01
                 ...         
504093    2020-12-31 23:00:00
504094    2020-12-31 23:00:00
504095    2020-12-31 23:00:00
504096    2020-12-31 23:00:00
504097    2020-12-31 23:00:00
Name: Fecha, Length: 504098, dtype: object

### Carga de fechas

Vuelve a cargar el dataset usando el parámetro `parse_dates = [0]`. Selecciona la columna `Fecha` del dataset. ¿De qué tipo es la columna `Fecha`? . 

In [72]:
parse_fechas = pd.read_csv('peatones_madrid/peatones_madrid.csv', delimiter = '|', parse_dates  = [0])
parse_fechas

Fecha     Hora    Identificador  Peatones  \
0      2019-01-01 00:00:00  0:00:00  PERM_PEA02_PM01     497.0   
1      2019-01-01 00:00:00  0:15:00  PERM_PEA02_PM01     244.0   
2      2019-01-01 00:00:00  0:30:00  PERM_PEA02_PM01     179.0   
3      2019-01-01 00:00:00  0:45:00  PERM_PEA02_PM01     174.0   
4      2019-01-01 00:00:00  1:00:00  PERM_PEA02_PM01     131.0   
...                    ...      ...              ...       ...   
504093 2020-12-31 23:00:00    23:00  PERM_PEA15_PM01      44.0   
504094 2020-12-31 23:00:00    23:00  PERM_PEA16_PM01      44.0   
504095 2020-12-31 23:00:00    23:00  PERM_PEA17_PM01     133.0   
504096 2020-12-31 23:00:00    23:00  PERM_PEA18_PM01     191.0   
504097 2020-12-31 23:00:00    23:00  PERM_PEA19_PM01      48.0   

        Número_Distrito     Distrito                   Nombre_Vial  Número  \
0                   1.0       Centro              Calle Fuencarral    22.0   
1                   1.0       Centro              Calle Fuencarral    22.0   
2                   1.0       Centro              Calle Fuencarral    22.0   
3                   1.0       Centro              Calle Fuencarral    22.0   
4                   1.0       Centro              Calle Fuencarral    22.0   
...                 ...          ...                           ...     ...   
504093              1.0       Centro              Calle Fuencarral    22.0   
504094              2.0   Arganzuela  Plaza del Emperador Carlos V    11.0   
504095              2.0   Arganzuela  Plaza del Emperador Carlos V    11.0   
504096              1.0       Centro              Calle Fuencarral    22.0   
504097              1.0       Centro              Calle Fuencarral    22.0   

        Código_Postal      Latitud      Longitud  
0             28004.0   40,4220090    -3,7008917  
1             28004.0   40,4220090    -3,7008917  
2             28004.0   40,4220090    -3,7008917  
3             28004.0   40,4220090    -3,7008917  
4             28004.0   40,4220090    -3,7008917  
...               ...          ...           ...  
504093        28004.0  40,42200905  -3,700891722  
504094        28012.0  40,41110888  -3,696471467  
504095        28012.0  40,41110888  -3,696471467  
504096        28004.0  40,42200905  -3,700891722  
504097        28004.0  40,42200905  -3,700891722  

[504098 rows x 11 columns]

In [73]:
parse_fechas.Fecha

0        2019-01-01 00:00:00
1        2019-01-01 00:00:00
2        2019-01-01 00:00:00
3        2019-01-01 00:00:00
4        2019-01-01 00:00:00
                 ...        
504093   2020-12-31 23:00:00
504094   2020-12-31 23:00:00
504095   2020-12-31 23:00:00
504096   2020-12-31 23:00:00
504097   2020-12-31 23:00:00
Name: Fecha, Length: 504098, dtype: datetime64[ns]

### Selección 2
Selecciona las columnas `Distrito`, `Peatones` y `Fecha`. El dataframe resultado ha de llamarse `temporal_distritos`.

In [74]:
temporal_distritos = peaton.loc[:,["Distrito","Peatones","Fecha"] ]
temporal_distritos

Distrito  Peatones                Fecha
0            Centro     497.0           2019-01-01
1            Centro     244.0           2019-01-01
2            Centro     179.0           2019-01-01
3            Centro     174.0           2019-01-01
4            Centro     131.0           2019-01-01
...             ...       ...                  ...
504093       Centro      44.0  2020-12-31 23:00:00
504094   Arganzuela      44.0  2020-12-31 23:00:00
504095   Arganzuela     133.0  2020-12-31 23:00:00
504096       Centro     191.0  2020-12-31 23:00:00
504097       Centro      48.0  2020-12-31 23:00:00

[504098 rows x 3 columns]

### Selección 3

Seleccionar las filas de la calle  `Calle Mayor`. La calle se registra bajo la columna `Nombre_Vial      `. El dataframe resultado ha de llamarse `df_calle_mayor`.

In [75]:
cond_mayor = peaton["Nombre_Vial"]== "Calle Mayor"
df_calle_mayor =peaton[cond_mayor]
df_calle_mayor

Fecha     Hora    Identificador  Peatones  \
16921            2019-01-01  0:00:00  PERM_PEA07_PM01      40.0   
16922            2019-01-01  0:15:00  PERM_PEA07_PM01      52.0   
16923            2019-01-01  0:30:00  PERM_PEA07_PM01      39.0   
16924            2019-01-01  0:45:00  PERM_PEA07_PM01      53.0   
16925            2019-01-01  1:00:00  PERM_PEA07_PM01      30.0   
...                     ...      ...              ...       ...   
376653  2020-09-30 19:00:00    19:00  PERM_PEA07_PM01     158.0   
376654  2020-09-30 20:00:00    20:00  PERM_PEA07_PM01     157.0   
376655  2020-09-30 21:00:00    21:00  PERM_PEA07_PM01     162.0   
376656  2020-09-30 22:00:00    22:00  PERM_PEA07_PM01      81.0   
376657  2020-09-30 23:00:00    23:00  PERM_PEA07_PM01      60.0   

        Número_Distrito Distrito  Nombre_Vial  Número  Código_Postal  \
16921               1.0   Centro  Calle Mayor     1.0        28013.0   
16922               1.0   Centro  Calle Mayor     1.0        28013.0   
16923               1.0   Centro  Calle Mayor     1.0        28013.0   
16924               1.0   Centro  Calle Mayor     1.0        28013.0   
16925               1.0   Centro  Calle Mayor     1.0        28013.0   
...                 ...      ...          ...     ...            ...   
376653              1.0   Centro  Calle Mayor    13.0        28013.0   
376654              1.0   Centro  Calle Mayor    13.0        28013.0   
376655              1.0   Centro  Calle Mayor    13.0        28013.0   
376656              1.0   Centro  Calle Mayor    13.0        28013.0   
376657              1.0   Centro  Calle Mayor    13.0        28013.0   

            Latitud      Longitud  
16921    40,4163760    -3,7060291  
16922    40,4163760    -3,7060291  
16923    40,4163760    -3,7060291  
16924    40,4163760    -3,7060291  
16925    40,4163760    -3,7060291  
...             ...           ...  
376653  40,41637598  -3,706029098  
376654  40,41637598  -3,706029098  
376655  40,41637598  -3,706029098  
376656  40,41637598  -3,706029098  
376657  40,41637598  -3,706029098  

[26865 rows x 11 columns]

### Selección 4
Seleccionar las filas del día 10 de Mayo de 2020 del código postal `28011`. El dataframe resultado ha de llamarse `df_10Mayo`.

In [76]:
a = parse_fechas.Código_Postal == 28011
fecha = parse_fechas.Fecha == "2020-05-10"
df_10Mayo = parse_fechas[a & fecha]
df_10Mayo

Fecha  Hora    Identificador  Peatones  Número_Distrito Distrito  \
463981 2020-05-10  0:00  PERM_PEA12_PM01      13.0             10.0   Latina   

                                              Nombre_Vial  Número  \
463981  Madrid Río. Puente de Segovia con Paseo Ermita...     1.0   

        Código_Postal      Latitud      Longitud  
463981        28011.0  40,41373441  -3,700891722

### Valores 1

Muestra un listado de los distintos distritos que hay en el dataset. Usa la operación `unique`. Observa los valores devueltos. 

In [77]:
parse_fechas.Distrito.unique()

array([' Centro', ' Chamberí', 'Latina', ' Moncloa-Aravaca',
       ' Arganzuela', 'Centro'], dtype=object)

### Map

Modifica el valor de la columna `Distrito` para que los valores aparezcan sin el espacio en blanco que aparece antes del nombre del distrito. Recuerda que la función `strip()` del tipo `str` se encarga de eliminar los espacios en blanco.
Muestra de nuevo el listado de los distintos distritos que hay en el dataset. Usa la operación `unique`. Observa los valores devueltos.

In [78]:
#no se como empezar

### Consulta 1
Calcular la media del número de peatones en el distrito `Centro`.

In [79]:
media = peaton.Distrito == "Centro"
peaton1 = peaton.Peatones
peaton1[Centro].mean()
# no se como corregir el error

NameError: name 'Centro' is not defined

### Valores 2

Deseamos conocer cuántas observaciones (filas) se han registrado en cada uno de los distritos. Usa `value_counts()`.

In [80]:
peaton.Distrito.value_counts()

 Centro             332590
 Arganzuela          53186
 Chamberí            51831
 Moncloa-Aravaca     27497
Latina               27296
Centro               11698
Name: Distrito, dtype: int64

### Consulta 2
¿Qué calle recoge el máximo número de peatones?

In [81]:
max = peaton[["Peatones","Nombre_Vial"]]
max.max()

Peatones                   50460
Nombre_Vial    Ronda de Valencia
dtype: object

### Consulta 3
Selecciona las filas correspondientes al identificador `PERM_PEA19_PM01` donde el número de peatones registrados sea 0.

In [82]:
identificador = peaton.Identificador == "PERM_PEA19_PM01"
reg = peaton.Peatones == 0
identificador1 = peaton[identificador & reg]
identificador1

Fecha  Hora    Identificador  Peatones  Número_Distrito  \
455576  2020-01-01 06:00:00  6:00  PERM_PEA19_PM01       0.0              1.0   
455596  2020-02-01 02:00:00  2:00  PERM_PEA19_PM01       0.0              1.0   
455600  2020-02-01 06:00:00  6:00  PERM_PEA19_PM01       0.0              1.0   
455624  2020-03-01 06:00:00  6:00  PERM_PEA19_PM01       0.0              1.0   
455645  2020-04-01 03:00:00  3:00  PERM_PEA19_PM01       0.0              1.0   
...                     ...   ...              ...       ...              ...   
501969  2020-12-27 07:00:00  7:00  PERM_PEA19_PM01       0.0              1.0   
503242  2020-12-30 02:00:00  2:00  PERM_PEA19_PM01       0.0              1.0   
503261  2020-12-30 03:00:00  3:00  PERM_PEA19_PM01       0.0              1.0   
503717  2020-12-31 03:00:00  3:00  PERM_PEA19_PM01       0.0              1.0   
503774  2020-12-31 06:00:00  6:00  PERM_PEA19_PM01       0.0              1.0   

       Distrito       Nombre_Vial  Número  Código_Postal      Latitud  \
455576   Centro      Calle Bailén    10.0        28013.0    40,415111   
455596   Centro      Calle Bailén    10.0        28013.0    40,415111   
455600   Centro      Calle Bailén    10.0        28013.0    40,415111   
455624   Centro      Calle Bailén    10.0        28013.0    40,415111   
455645   Centro      Calle Bailén    10.0        28013.0    40,415111   
...         ...               ...     ...            ...          ...   
501969   Centro  Calle Fuencarral    22.0        28004.0  40,42200905   
503242   Centro  Calle Fuencarral    22.0        28004.0  40,42200905   
503261   Centro  Calle Fuencarral    22.0        28004.0  40,42200905   
503717   Centro  Calle Fuencarral    22.0        28004.0  40,42200905   
503774   Centro  Calle Fuencarral    22.0        28004.0  40,42200905   

            Longitud  
455576     -3,713569  
455596     -3,713569  
455600     -3,713569  
455624     -3,713569  
455645     -3,713569  
...              ...  
501969  -3,700891722  
503242  -3,700891722  
503261  -3,700891722  
503717  -3,700891722  
503774  -3,700891722  

[581 rows x 11 columns]

### Modificar columnas 1
Crea una nueva columna llamada 'Peatones 1000', que represente el número de peatones expresado en miles. Es decir, para el valor de número de peatones 50460, la nueva columna debe tener el valor 50.460.

In [83]:
peaton["Peatones 1000"] = peaton.Peatones/1000
peaton

Fecha     Hora    Identificador  Peatones  \
0                2019-01-01  0:00:00  PERM_PEA02_PM01     497.0   
1                2019-01-01  0:15:00  PERM_PEA02_PM01     244.0   
2                2019-01-01  0:30:00  PERM_PEA02_PM01     179.0   
3                2019-01-01  0:45:00  PERM_PEA02_PM01     174.0   
4                2019-01-01  1:00:00  PERM_PEA02_PM01     131.0   
...                     ...      ...              ...       ...   
504093  2020-12-31 23:00:00    23:00  PERM_PEA15_PM01      44.0   
504094  2020-12-31 23:00:00    23:00  PERM_PEA16_PM01      44.0   
504095  2020-12-31 23:00:00    23:00  PERM_PEA17_PM01     133.0   
504096  2020-12-31 23:00:00    23:00  PERM_PEA18_PM01     191.0   
504097  2020-12-31 23:00:00    23:00  PERM_PEA19_PM01      48.0   

        Número_Distrito     Distrito                   Nombre_Vial  Número  \
0                   1.0       Centro              Calle Fuencarral    22.0   
1                   1.0       Centro              Calle Fuencarral    22.0   
2                   1.0       Centro              Calle Fuencarral    22.0   
3                   1.0       Centro              Calle Fuencarral    22.0   
4                   1.0       Centro              Calle Fuencarral    22.0   
...                 ...          ...                           ...     ...   
504093              1.0       Centro              Calle Fuencarral    22.0   
504094              2.0   Arganzuela  Plaza del Emperador Carlos V    11.0   
504095              2.0   Arganzuela  Plaza del Emperador Carlos V    11.0   
504096              1.0       Centro              Calle Fuencarral    22.0   
504097              1.0       Centro              Calle Fuencarral    22.0   

        Código_Postal      Latitud      Longitud  Peatones 1000  
0             28004.0   40,4220090    -3,7008917          0.497  
1             28004.0   40,4220090    -3,7008917          0.244  
2             28004.0   40,4220090    -3,7008917          0.179  
3             28004.0   40,4220090    -3,7008917          0.174  
4             28004.0   40,4220090    -3,7008917          0.131  
...               ...          ...           ...            ...  
504093        28004.0  40,42200905  -3,700891722          0.044  
504094        28012.0  40,41110888  -3,696471467          0.044  
504095        28012.0  40,41110888  -3,696471467          0.133  
504096        28004.0  40,42200905  -3,700891722          0.191  
504097        28004.0  40,42200905  -3,700891722          0.048  

[504098 rows x 12 columns]

### Modificar columnas 2
El tipo de la columna `Latitud` debería ser `float`. Modifica dicha columna para que su tipo sea `float`.

In [84]:
lat = peaton.Latitud.apply(lambda x: x.replace(",","."))
lat

0          40.4220090
1          40.4220090
2          40.4220090
3          40.4220090
4          40.4220090
             ...     
504093    40.42200905
504094    40.41110888
504095    40.41110888
504096    40.42200905
504097    40.42200905
Name: Latitud, Length: 504098, dtype: object

In [85]:
lat.astype(float)

0         40.422009
1         40.422009
2         40.422009
3         40.422009
4         40.422009
            ...    
504093    40.422009
504094    40.411109
504095    40.411109
504096    40.422009
504097    40.422009
Name: Latitud, Length: 504098, dtype: float64